rust/types/src/regex/literal_expr.rs

```diff
@@ -115,95 +185,179 @@ pub trait NgramLiteralProvider<E, const N: usize = 3> {
+                let suffix = &ngram[ngram.char_indices().nth(1).unwrap_or_default().0..];
+                lookup_table
+                    .prefix
+                    .entry(prefix)
+                    .or_insert_with(|| Vec::with_capacity(ngrams.len()))
+                    .push(ngram_doc_pos_index);
+                lookup_table
+                    .suffix
+                    .entry(suffix)
+                    .or_insert_with(|| Vec::with_capacity(ngrams.len()))
+                    .push(ngram_doc_pos_index);
+            }
+            let lookup_table_index = lookup_table_vec.len();
+            lookup_table_vec.push(lookup_table);
+            if lookup_table_size < min_lookup_table_size {
+                min_lookup_table_size = lookup_table_size;
+                min_lookup_table_index = lookup_table_index;
             }
         }
 
-        for literal in remaining_literals {
-            if suffix_doc_pos.is_empty() {
-                break;
-            }
-            let mut new_suffix_doc_pos: HashMap<Vec<char>, HashMap<u32, HashSet<u32>>> =
-                HashMap::new();
-            for (mut suffix, doc_pos) in suffix_doc_pos {
-                let ngram_ranges = match literal {
-                    Literal::Char(literal_char) => {
-                        suffix.push(*literal_char);
-                        vec![(suffix.clone(), suffix)]
+        // Gather candidate documents
+        let min_lookup_table = &lookup_table_vec[min_lookup_table_index];
+        let min_ngram_doc_pos_iter = min_lookup_table
+            .prefix
+            .values()
+            .flat_map(|idxs| idxs.iter().map(|idx| &ngram_doc_pos_vec[*idx]));
+        let mut candidates =
+            HashMap::<_, Vec<_>>::with_capacity(min_ngram_doc_pos_iter.clone().map(Vec::len).sum());
+        for (ngram, doc, pos) in min_ngram_doc_pos_iter
+            .flatten()
+            .filter(|(_, d, _)| mask.is_none() || mask.is_some_and(|m| m.contains(d)))
+        {
+            candidates
+                .entry(*doc)
+                .or_insert_with(|| Vec::with_capacity(min_lookup_table.prefix.len()))
+                .push((*ngram, *pos));
+        }
+
+        // Find a valid trace across lookup tables
+        let mut result = HashSet::with_capacity(candidates.len());
+        for (doc, pivot_ngram_pos) in candidates {
+            for (ngram, pos) in pivot_ngram_pos
+                .into_iter()
+                .flat_map(|(n, ps)| ps.iter().map(move |p| (n, *p)))
+            {
+                // Trace to the right of pivot
+                let mut suffix_pos_idx =
+                    Vec::with_capacity(lookup_table_vec.len() - min_lookup_table_index);
+                let suffix_offset = ngram.char_indices().nth(1).unwrap_or_default().0;
+                suffix_pos_idx.push((&ngram[suffix_offset..], pos + suffix_offset as u32, 0));
+                while let Some((suffix, match_pos, ngram_index)) = suffix_pos_idx.pop() {
```

Comment:
worth adding comments here. this is an iterative DFS right? Maybe a recursive impl is more readable